sources/System/Lobby/ResumeLobbySystem.cpp

```diff
@@ -0,0 +1,226 @@
+#include <Component/Position/PositionComponent.hpp>
+#include <Component/Renderer/Drawable3DComponent.hpp>
+#include <Map/Map.hpp>
+#include <Component/BombHolder/BombHolderComponent.hpp>
+#include <Parser/ParserYaml.hpp>
+#include <Drawables/2D/Text.hpp>
+#include "Component/Color/ColorComponent.hpp"
+#include "Component/Stat/StatComponent.hpp"
+#include "Component/Bonus/PlayerBonusComponent.hpp"
+
+namespace RAY3D = RAY::Drawables::Drawables3D;
+namespace RAY2D = RAY::Drawables::Drawables2D;
+
+namespace BBM
+{
+	std::array<std::string, 4> ResumeLobbySystem::_colors = {
+			"blue",
+			"red",
+			"green",
+			"yellow"
+	};
+
+	std::array<RAY::Color, 4> ResumeLobbySystem::_rayColors = {
+			BLUE,
+			RED,
+			GREEN,
+			YELLOW
+	};
+
+	ResumeLobbySystem::ResumeLobbySystem(WAL::Wal &wal)
+			: System(wal)
+	{}
+
+	void ResumeLobbySystem::onUpdate(WAL::ViewEntity<ResumeLobbyComponent, Drawable2DComponent> &entity, std::chrono::nanoseconds dtime)
+	{
+		auto &lobby = entity.get<ResumeLobbyComponent>();
+
+		auto lastTick = std::chrono::steady_clock::now();
+		if (lastTick - lobby.lastInput < std::chrono::milliseconds(150))
+			return;
+
+		if (lobby.layout == ControllableComponent::NONE) {
+			for (auto &[_, ctrl] : this->_wal.getScene()->view<ControllableComponent>()) {
+				auto &controller = ctrl;
+				if (controller.select) {
+					if (std::any_of(this->getView().begin(), this->getView().end(), [&controller](WAL::ViewEntity<ResumeLobbyComponent, Drawable2DComponent> &view) {
+						return view.get<ResumeLobbyComponent>().layout == controller.layout;
+					}))
+						return;
+					lobby.ready = true;
+					lobby.lastInput = lastTick;
+					lobby.layout = controller.layout;
+					controller.select = false;
+					this->_wal.getSystem<MenuControllableSystem>().now = lastTick;
+					auto *texture = dynamic_cast<RAY::Texture *>(lobby.readyButton.getComponent<Drawable2DComponent>().drawable.get());
+					if (texture)
+						texture->use("assets/player/icons/ready.png");
+					return;
+				}
+			}
+		}
+	}
+
+	void ResumeLobbySystem::onSelfUpdate(std::chrono::nanoseconds dtime)
+	{
+		auto &view = this->_wal.getScene()->view<TagComponent<"PlayButton">, Drawable2DComponent>();
+		if (view.size() == 0)
+			return;
+		auto *texture = dynamic_cast<RAY::Texture *>(view.front().get<Drawable2DComponent>().drawable.get());
+		if (!texture)
+			return;
+		if (playersAreReady(*this->_wal.getScene()))
+			texture->setColor(WHITE);
+		else
+			texture->setColor(GRAY);
+	}
+
+	bool ResumeLobbySystem::playersAreReady(WAL::Scene &scene)
+	{
+		auto &lobby = scene.view<ResumeLobbyComponent>();
+		return std::all_of(lobby.begin(), lobby.end(), [](WAL::ViewEntity<ResumeLobbyComponent> &entity) {
+			auto &lobbyPlayer = entity.get<ResumeLobbyComponent>();
+			return lobbyPlayer.ready && lobbyPlayer.layout != ControllableComponent::NONE;
+		});
+	}
+
+	void ResumeLobbySystem::addController(WAL::Entity &player, ControllableComponent::Layout layout)
```

Comment:
Don't copy paste this, the method is public static. You can simply use the lobby method (+ it has been modified since to support replay)